src/transformers/generation_beam_constraints.py

```diff
@@ -202,6 +211,170 @@ def copy(self, stateful=False):
+
+        return next_tokens
+
+    def reached_leaf(self, current_seq):
+        next_tokens = self.next_tokens(current_seq)
+
+        return -1 in next_tokens
+
+
+class DisjunctiveConstraint(Constraint):
+    r"""
+    A special [`Constraint`] that is fulfilled by fulfilling just one of several constraints.
+
+    Args:
+        constraints (`List[Constraint]`):
+            The list of [`Constraint`] objects, fulfillment of just one of which fulfills this
+            [`DisjunctiveConstraint`].
+    """
+
+    def __init__(self, nested_token_ids: List[Constraint]):
+        super(Constraint, self).__init__()
+
+        self.token_ids = nested_token_ids
+        self._check_valid(self.token_ids)
+
+        self.trie = DisjunctiveTrie([ids.tolist() for ids in self.token_ids])
+
+        self.seqlen = self.trie.max_height
+        self.current_seq = []
+        self.completed = False
+
+    def _check_valid(self, nested_token_ids):
+        """
+        Checks whether nested_token_ids is valid.
+        1. Checks Type
+        2. Checks for illegal subsets
+            We're checking if there is a `seq_k` in `nested_token_ids` such that `seq_k` is 100% a subset found in some
+            seq in `nested_token_ids`. We can't allow this because it leads to a perverse interpretation of
+            "completeness". Luckily, we only need to check the beginning of the sequence since it's a trie
+            representation.
+        """
+        if not isinstance(nested_token_ids, list):
+            nested_token_ids = nested_token_ids.tolist()
+            self.token_ids = nested_token_ids
+
+        for k, seq_k in enumerate(nested_token_ids):
+            is_tensor = isinstance(seq_k, torch.Tensor)
+            is_int_list = isinstance(seq_k, List) and isinstance(seq_k[0], int)
+            is_int_tensor = is_single_int_tensor(seq_k)
+
+            not_positive = torch.any(seq_k < 0) if is_tensor else len([t for t in seq_k if t < 0]) > 0
+            if isinstance(seq_k, int) or not (is_int_list or is_int_tensor) or not_positive:
+                raise ValueError(
+                    f"Sequences within `nested_token_ids` has to be a single list or tensor of positive integers but is {seq_k}"
+                )
+
+            seq_k = seq_k.tolist() if is_int_tensor else seq_k
+            for i, seq_i in enumerate(nested_token_ids[k + 1 :]):
+                seq_i = seq_i.tolist() if is_single_int_tensor(seq_i) else seq_i
+
+                if seq_k == seq_i[: len(seq_k)] or seq_k[: len(seq_i)] == seq_i:
+                    raise ValueError(
+                        "No sequence in `nested_token_ids` can be a complete subset of another, but found"
+                        f" sequences {seq_k} and {seq_i}."
+                    )
+
+            self.token_ids[k] = torch.tensor(seq_k)
+
+        return True
+
+    def advance(self):
+        token_list = self.trie.next_tokens(self.current_seq)
+        token_list = [t for t in token_list if t >= 0]
```

Comment:
This seems to actually just ignore subset options.

Reply:
Right. To be more exact, after the checking of subsets in `__init__`, the rest of the code can safely assume there are no subset options.